File: Algorithms/C++/Stack/20_Valid_Parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> stk;

        for (int i = 0; i < s.length(); i++) {
            char c = s[i];

            if (c == '(' || c == '{' || c == '[') {
                stk.push(c);
            } else {
                if (stk.empty()) {
                    return false;
                }
                char top = stk.top();
                if (c == ')' && top != '('
                || c == '}' && top != '{'
                || c == ']' && top != '[') {
                    return false;
                }
                stk.pop();
            }
        }

        if (!stk.empty()) {
            return false;
        }
        return true;
    }
};

File: Algorithms/Swift/Stack/20_Valid_Parentheses.swift
class Solution {
    func isValid(s: String) -> Bool {
        if s.isEmpty {
            return true
        }

        var stack = [Character]()

        for c in s.characters {
            switch c {
            case "(", "{", "[":
                stack.append(c);
            case ")":
                let bracket = stack.last
                if bracket == nil || bracket! != "(" {
                    return false
                } else {
                    stack.removeLast()
                }
            case "}":
                let bracket = stack.last
                if bracket == nil || bracket! != "{" {
                    return false
                } else {
                    stack.removeLast()
                }
            case "]":
                let bracket = stack.last
                if bracket == nil || bracket! != "[" {
                    return false
                } else {
                    stack.removeLast()
                }
            default:
                break
            }
        }

        return stack.isEmpty
    }
}
